kmip/services/server/crypto/engine.py

```diff
@@ -148,6 +156,71 @@ def create_asymmetric_key_pair(self, algorithm, length):
         engine_method = self._asymetric_key_algorithms.get(algorithm)
         return engine_method(length)
 
+    def mac(self, algorithm, key, data):
+        """
+        Generate message authentication code.
+
+        Args:
+            algorithm(CryptographicAlgorithm): An enumeration specifying the
+                algorithm for which the MAC operation will use.
+            key(bytes): secret key used in the MAC operation
+            data(bytes): The data to be MACed.
+
+        Returns:
+            bytes: The MACed data
+
+        Raises:
+            InvalidField: Raised when the algorithm is unsupported or the
+                length is incompatible with the algorithm.
+            CryptographicFailure: Raised when the key generation process
+                fails.
+
+        Example:
+            >>> engine = CryptographyEngine()
+            >>> mac_data = engine.mac(
+            ...     CryptographicAlgorithm.HMAC-SHA256, b'\x01\x02\x03\x04')
+        """
+
+        mac_data = None
+
+        if algorithm in self._hash_algorithms.keys():
+            self.logger.info(
+                "Generating hash-based Message authentication codes using {0}".
+                format(algorithm.name)
+            )
+            hash_algorithm = self._hash_algorithms.get(algorithm)
+            try:
+                h = hmac.HMAC(key, hash_algorithm(), backend=default_backend())
+                h.update(data)
+                mac_data = h.finalize()
+            except Exception as e:
+                self.logger.exception(e)
+                raise exceptions.CryptographicFailure(
+                    "An error occurred while doing HMAC operation. "
+                    "See the server log for more information."
+                )
+        elif algorithm in self._symmetric_key_algorithms.keys():
+            self.logger.info(
+                "Generating cipher-based Message authentication codes using "
+                "{0}".format(algorithm.name)
+            )
+            cipher_algorithm = self._symmetric_key_algorithms.get(algorithm)
+            try:
+                c = cmac.CMAC(cipher_algorithm(key), backend=default_backend())
```

Comment:
Nice job here including CMAC in addition to HMAC. The KMIP specification doesn't indicate what specific type of MAC operation to support, so anything possible in the list of CryptographicAlgorithms is valid.

We should probably add a comment here indicating that the ARC4 and IDEA algorithms aren't compatible with CMAC; CMAC requires block ciphers and those two aren't block ciphers.